cdap-kafka/src/main/java/co/cask/cdap/kafka/run/KafkaServerMain.java

```diff
@@ -102,11 +94,25 @@ public void init(String[] args) {
       }
     }
 
-    int brokerId = generateBrokerId(address);
-    LOG.info(String.format("Initializing server with broker id %d", brokerId));
+    kafkaProperties = generateKafkaConfig(cConf);
 
-    kafkaProperties = generateKafkaConfig(brokerId, zkConnectStr, hostname, port, numPartitions,
-                                          replicationFactor, logDir);
+    Preconditions.checkState(Integer.parseInt(
+      kafkaProperties.getProperty(KafkaConstants.ConfigKeys.NUM_PARTITIONS_CONFIG)) > 0,
+                             "Num partitions should be greater than zero.");
+    int port = Integer.getInteger(kafkaProperties.getProperty(KafkaConstants.ConfigKeys.PORT_CONFIG), -1);
+    Preconditions.checkState(port > 0, "Port number is invalid.");
+
+    if (kafkaProperties.get("broker.id") == null) {
+      int brokerId = generateBrokerId(address);
+      LOG.info(String.format("Initializing server with broker id %d", brokerId));
+      kafkaProperties.setProperty("broker.id", Integer.toString(brokerId));
+    }
+    if (hostname != null) {
```

Comment:
Kafka documentation for `host.name`: 
```
DEPRECATED: only used when `listeners` is not set. Use `listeners` instead. hostname of broker. If this is set, it will only bind to this address. If this is not set, it will bind to all interfaces
```
Should I change the config key to be listeners then?